### PR TITLE
Added backup verification before restoring it in test case PXBackupEndToEndBackupAndRestoreWithUpgrade

### PR DESCRIPTION
**Pull request by snigdha-px**
**What this PR does / why we need it**:
Added backup verification before restoring it in test case PXBackupEndToEndBackupAndRestoreWithUpgrade
**Which issue(s) this PR fixes** (optional)
Closes #https://portworx.atlassian.net/browse/PA-1771

**Special notes for your reviewer**:
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/upgrade-tests/job/Px-Backup-with-Stork-Upgrade/job/Px-Backup+Stork-Upgrade-Vanilla/183/consoleFull

### CODE REVIEW

**Review by kshithijiyer-px — APPROVED** [review]
LGTM

File: tests/backup/backup_upgrade_test.go [hunk 1]
```diff
@@ -612,6 +612,9 @@ var _ = Describe("{PXBackupEndToEndBackupAndRestoreWithUpgrade}", func() {
 			restoreSingleNSBackupInVariousWaysTask := func(index int, namespace string) {
 				nextScheduleBackupName, err := GetNextScheduleBackupName(singleNSScheduleNames[index], time.Duration(intervalInMins), ctx)
 				dash.VerifyFatal(err, nil, fmt.Sprintf("Fetching next schedule backup name of schedule named [%s]", singleNSScheduleNames[index]))
+				appContextsToBackup := FilterAppContextsByNamespace(srcClusterContexts, []string{namespace})
+				err = backupSuccessCheckWithValidation(ctx, nextScheduleBackupName, appContextsToBackup, orgID, maxWaitPeriodForBackupCompletionInMinutes*time.Minute, 30*time.Second)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verification of success of next single namespace schedule backup [%s] of schedule %s", nextScheduleBackupName, singleNSScheduleNames[index]))
```
Comment (kphalgun-px) [inline tests/backup/backup_upgrade_test.go:617]: Can we replace CreateBackup with CreateBackupWithValidation?

Comment (kphalgun-px) [reply tests/backup/backup_upgrade_test.go:617]: and CreateRestore with CreateRestoreWithValidation?

**Review by kphalgun-px — APPROVED** [review]
LGTM

File: tests/backup/backup_upgrade_test.go (hunk 1 above)
Comment (snigdha-px) [reply tests/backup/backup_upgrade_test.go:617]: At which line no are you asking to use createBackupWithValidation?
@kphalgun-px

Comment (kphalgun-px) [reply tests/backup/backup_upgrade_test.go:617]: Everywhere in the test case.

Comment (snigdha-px) [reply tests/backup/backup_upgrade_test.go:617]: Yes will change that, but we still will need to use backupSuccessCheckWithValidation for next schedule backups

Comment (kphalgun-px) [reply tests/backup/backup_upgrade_test.go:617]: Right. My bad.

Comment (snigdha-px) [reply tests/backup/backup_upgrade_test.go:622]: @kphalgun-px 
changed CreateRestore with CreateRestoreWithValidation and CreateBackup with CreateBackupWithValidation

File: tests/backup/backup_upgrade_test.go [hunk 2]
```diff
@@ -248,7 +248,7 @@ var _ = Describe("{PXBackupEndToEndBackupAndRestoreWithUpgrade}", func() {
 		restoreNames                       []string
 		mutex                              sync.Mutex
 	)
-
+	var backupToContextMapping = make(map[string][]*scheduler.Context, 0)
```
Comment (kphalgun-px) [inline tests/backup/backup_upgrade_test.go:251]: Can we put this along with other variables?

File: tests/backup/backup_upgrade_test.go [hunk 3]
```diff
@@ -393,23 +393,25 @@ var _ = Describe("{PXBackupEndToEndBackupAndRestoreWithUpgrade}", func() {
...
 				log.InfoD("Creating a backup of namespace [%s] with pre and post exec rules", namespace)
-				err := CreateBackup(backupName, SourceClusterName, backupLocationName, backupLocationUid, []string{namespace},
-					labelSelectors, orgID, srcClusterUid, preRuleNames[appName], preRuleUids[appName], postRuleNames[appName], postRuleUids[appName], ctx)
+				appContextsToBackup := FilterAppContextsByNamespace(srcClusterContexts, []string{namespace})
+				err = CreateBackupWithValidation(ctx, backupName, SourceClusterName, backupLocationName, backupLocationUid, appContextsToBackup, labelSelectors, orgID, srcClusterUid, preRuleNames[appName], preRuleUids[appName], postRuleNames[appName], postRuleUids[appName])
 				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of backup [%s]", backupName))
 				err = IsFullBackup(backupName, orgID, ctx)
 				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying if backup [%s] is a full backup", backupName))
 				backupWithRuleNames = SafeAppend(&mutex, backupWithRuleNames, backupName).([]string)
+				backupToContextMapping[backupName] = appContextsToBackup
```
Comment (kphalgun-px) [inline tests/backup/backup_upgrade_test.go:402]: How about using a mutex to be safe?

An anonymous function

```
updatebackupToContextMapping := func(backupName string, appContextsToBackup string) {
	mutex.Lock()
	defer mutex.Unlock()
	backupMapping[backupName] = appContextsToBackup
}
```

and

```
updatebackupToContextMapping(backupName, appContextsToBackup)
```

File: tests/backup/backup_upgrade_test.go (hunk 2 above)
Comment (snigdha-px) [reply tests/backup/backup_upgrade_test.go:251]: done

File: tests/backup/backup_upgrade_test.go [hunk 4]
```diff
@@ -642,7 +650,8 @@ var _ = Describe("{PXBackupEndToEndBackupAndRestoreWithUpgrade}", func() {
 					restoreName := fmt.Sprintf("%s-%s", config.namePrefix, RandomString(4))
 					log.InfoD("Restoring next single namespace schedule backup [%s] in cluster [%s] with restore [%s] and namespace mapping %v", nextScheduleBackupName, destinationClusterName, restoreName, config.namespaceMapping)
 					if config.replacePolicy == ReplacePolicy_Retain {
-						err = CreateRestore(restoreName, nextScheduleBackupName, config.namespaceMapping, destinationClusterName, orgID, ctx, config.storageClassMapping)
+						appContextsToBackup := FilterAppContextsByNamespace(srcClusterContexts, []string{namespace})
+						err = CreateRestoreWithValidation(ctx, restoreName, nextScheduleBackupName, config.namespaceMapping, config.storageClassMapping, destinationClusterName, orgID, appContextsToBackup)
```
Comment (kphalgun-px) [inline tests/backup/backup_upgrade_test.go:654]: I do not think we can run `CreateRestoreWithValidation` in parallel. I apologize. Can you please replace the Parallel parameter with Sequential in TaskHandler if it processes `CreateRestoreWithValidation`? I also hope that this test case does not take a backup from the destination cluster.

File: tests/backup/backup_upgrade_test.go (hunk 1 above)
Comment (kphalgun-px) [reply tests/backup/backup_upgrade_test.go:622]: Thank you.

File: tests/backup/backup_upgrade_test.go (hunk 3 above)
Comment (snigdha-px) [reply tests/backup/backup_upgrade_test.go]: done

File: tests/backup/backup_upgrade_test.go (hunk 4 above)
Comment (snigdha-px) [reply tests/backup/backup_upgrade_test.go:663]: done

**Review by mkoppal-px — APPROVED** [review]
LGTM

**Pull request merged**